File: PHP/src/birthday_greetings/BirthdayService.php
<?php

namespace birthday_greetings;

use PHPMailer\PHPMailer\PHPMailer;
use PHPMailer\PHPMailer\Exception;

class BirthdayService
{
    public function sendGreetings(string $fileName, XDate $xDate, string $smtpHost, int $smtpPort): void
    {
        $handle = fopen($fileName, 'r');
        if ($handle === false) {
            throw new \RuntimeException("Could not open file: $fileName");
        }

        // Skip header
        fgetcsv($handle, 0, ',', '"', '\\');

        while (($data = fgetcsv($handle, 0, ',', '"', '\\')) !== false) {
            if (count($data) < 4) {
                continue; // Skip invalid lines
            }

            try {
                $employee = new Employee(trim($data[1]), trim($data[0]), trim($data[2]), trim($data[3]));
                if ($employee->isBirthday($xDate)) {
                    $recipient = $employee->getEmail();
                    $body = str_replace('%NAME%', $employee->getFirstName(), 'Happy Birthday, dear %NAME%');
                    $subject = 'Happy Birthday!';
                    $this->sendMessage($smtpHost, $smtpPort, 'sender@here.com', $subject, $body, $recipient);
                }
            } catch (\Exception $e) {
                // Log error and continue with next employee
                error_log("Error processing employee data: " . $e->getMessage());
                continue;
            }
        }

        fclose($handle);
    }

    private function sendMessage(string $smtpHost, int $smtpPort, string $sender, string $subject, string $body, string $recipient): void
    {
        $mail = new PHPMailer(true);

        try {
            // Server settings
            $mail->isSMTP();
            $mail->Host = $smtpHost;
            $mail->Port = $smtpPort;
            $mail->SMTPAuth = false;

            // Recipients
            $mail->setFrom($sender);
            $mail->addAddress($recipient);

            // Content
            $mail->isHTML(false);
            $mail->Subject = $subject;
            $mail->Body = $body;

            $mail->send();
        } catch (Exception $e) {
            throw new \RuntimeException("Message could not be sent. Mailer Error: {$mail->ErrorInfo}", 0, $e);
        }
    }
}

File: Go/pkg/birthday_greetings/birthday_service.go
package birthday_greetings

import (
	"bufio"
	"fmt"
	"net/smtp"
	"os"
	"strings"
)

// BirthdayService handles sending birthday greetings
type BirthdayService struct{}

// NewBirthdayService creates a new BirthdayService
func NewBirthdayService() *BirthdayService {
	return &BirthdayService{}
}

// SendGreetings sends birthday greetings to employees whose birthday is today
func (bs *BirthdayService) SendGreetings(fileName string, xDate *XDate, smtpHost string, smtpPort int) error {
	file, err := os.Open(fileName)
	if err != nil {
		return fmt.Errorf("failed to open file: %w", err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	// Skip header
	if scanner.Scan() {
		_ = scanner.Text()
	}

	for scanner.Scan() {
		line := scanner.Text()
		employeeData := strings.Split(line, ", ")

		if len(employeeData) < 4 {
			continue // Skip invalid lines
		}

		employee, err := NewEmployee(employeeData[1], employeeData[0], employeeData[2], employeeData[3])
		if err != nil {
			return fmt.Errorf("failed to create employee: %w", err)
		}

		if employee.IsBirthday(xDate) {
			recipient := employee.GetEmail()
			body := strings.Replace("Happy Birthday, dear %NAME%", "%NAME%", employee.GetFirstName(), -1)
			subject := "Happy Birthday"
			err := bs.sendMessage(smtpHost, smtpPort, "sender@here.com", subject, body, recipient)
			if err != nil {
				return fmt.Errorf("failed to send message: %w", err)
			}
		}
	}

	if err := scanner.Err(); err != nil {
		return fmt.Errorf("error reading file: %w", err)
	}

	return nil
}

// sendMessage sends an email message
func (bs *BirthdayService) sendMessage(smtpHost string, smtpPort int, sender, subject, body, recipient string) error {
	// Construct the message
	message := []byte(fmt.Sprintf("From: %s\r\n"+
		"To: %s\r\n"+
		"Subject: %s\r\n"+
		"\r\n"+
		"%s\r\n", sender, recipient, subject, body))

	// Connect to the server, authenticate, and send the message
	addr := fmt.Sprintf("%s:%d", smtpHost, smtpPort)
	auth := smtp.PlainAuth("", "", "", smtpHost)

	// In a real application, we would use authentication, but for simplicity and to match
	// the Java version's behavior, we'll skip it for localhost
	var err error
	if smtpHost == "localhost" {
		err = smtp.SendMail(addr, nil, sender, []string{recipient}, message)
	} else {
		err = smtp.SendMail(addr, auth, sender, []string{recipient}, message)
	}

	if err != nil {
		return fmt.Errorf("failed to send email: %w", err)
	}

	return nil
}
